### PR TITLE
198 different folder names

### PR DESCRIPTION
**Pull request by erikbra**
Implemented fully configurable folders, both specifying custom folder names for the default folders (beforeMigration, up, afterMigration, etc), and specifying a completely custom folder structure.

Had to rewrite the core logic og GrateMigrator a bit to cater for different folders.

Could someone (@wokket , @DjKillerMemeStar, or others, of course,) please have a quick look at the new feature, and check if the documentation of the new feature solves #198 as intended, and that the specification of folder names on the command line are sensible to someone other than myself as well?

This is the documentation for the new folder configuration:

https://github.com/erikbra/grate/blob/5feb5694cfb48871aee78a16a2670b36119fe2e1/docs/ConfigurationOptions/FolderConfiguration.md

### CODE REVIEW

**Comment by wokket** [comment]
Fair bit to review there mate ;)  I'll try and get time to look at it tonight or tomorrow 👍

**Comment by erikbra** [comment]
Thanks a lot, @wokket , don't stress it, I mainly want a sanity check on whether the feature as documented makes sense, you don't have to dig into all the refactoring of code (unless you want to, of course) :)

**Comment by DjKillerMemeStar** [comment]
I'll also see what i can do in the following days!

**Comment by erikbra** [comment]
> I'll also see what i can do in the following days!

Thanks a lot, @DjKillerMemeStar! I appreciate it!

File: docs/ConfigurationOptions/FolderConfiguration.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+| ------ | ------- |------- |
+| One-time scripts | These are scripts that are run **exactly once** per database, and never again. | [One time scripts](../ScriptTypes/OneTimeScripts.md#one-time-scripts) |
+| Anytime Scripts | These scripts are run **any time they're changed** | [Anytime scripts](../ScriptTypes/AnytimeScripts.md#anytime-scripts) |
+| Everytime Scripts | These scripts are run (you guessed it) **every time** grate executes :) |  [Everytime scripts](../ScriptTypes/EverytimeScripts.md#everytime-scripts) |
+
+## Specifying a custom folder configuration
+
+If you want to specify a custom folder configuration, you have two options:
+
+1. Start with the [default grate folder configuration](#default-folder-configuration), and make adjusments
```
Comment (wokket) [inline docs/ConfigurationOptions/FolderConfiguration.md:31]: adjus**t**ments

File: docs/ConfigurationOptions/FolderConfiguration.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+nav_order: 5
+has_children: true
+---
+
+# Folder configuration
+
+The folder configuration in grate is now (since grate v 1.4) entirely configurable. That means you can 
+specify the folder structure of choice, and make grate work with your existing structure, rather than
+having to change the folder structure to suite how grate works. There is, however, a [default configuration](#default-folder-configuration),
+which might be a good starting point if you have no special requirements.
```
Comment (wokket) [inline docs/ConfigurationOptions/FolderConfiguration.md:14]: Given the complexity (and some subtlety called out in your 'interprestion' note below, could I suggest we reword this as the default config "...provides the simplest starting point..." or similar to really try and dissuade people from shooting themselves in the foot with this unless they need it?

File: docs/GettingStarted.md [hunk 3]
```diff
@@ -45,7 +45,10 @@ These scripts are run (you guessed it) **every time** grate executes :)  More in
 
 ## Directory run order
 
-grate processes the files in a standard set of directories in a fixed order for deterministic processing.  Folders are run in the following order:
+By default, grate processes the files in a standard set of directories in a fixed order for deterministic processing.  
+As of grate 1.4, the folder structure is [fully customisable](ConfigurationOptions/FolderConfiguration.md).
```
Comment (wokket) [inline docs/GettingStarted.md:49]: Again, maybe a gentle nudge here to stick with the defaults unless you know you need otherwise?

File: docs/ConfigurationOptions/FolderConfiguration.md (hunk 2 above)
Comment (erikbra) [reply docs/ConfigurationOptions/FolderConfiguration.md:14]: Just to clarify, @wokket, would you prefer the wording:

There is, however, a [default configuration](#default-folder-configuration),
which provides the simplest starting point if you have no special requirements.

And, is the feature itself too complex, do you think? Could it be made easier to use?

Comment (wokket) [reply docs/ConfigurationOptions/FolderConfiguration.md:14]: That wording sounds good to me.  I don't think it's _too_ complex, bit it's definitely _more_ complex than using the defaults.

I spend a bit of time in the MassTransit discord, and a surprisingly large amount of support issues boil down to "There was a knob, so I twiddled it, and now stuff's broken"... the solution to which is always "leave the knobs alone unless you **know** you need them".

I'm just trying to cut off similar issues here early by keeping people new to the tool on the simplest path possible.

Comment (DjKillerMemeStar) [reply docs/ConfigurationOptions/FolderConfiguration.md:14]: Just to pitch in, maybe we can offset @wokket's "issue" aswell by having the folderConfiguration in an advanced configurations. This way normal users are less likely to come across it.

Regarding the rest of the PR, it looks good to me!
